### PR TITLE
[Rate limit processor] Add counter metric for dropped events

### LINKED ISSUE #21020: Add event rate quota per Cloud Foundry organization
**Describe the enhancement:**

Add some kind of event rate quota per Cloud Foundry organization, to allow to dropping events once some limit is reached.

There is an issue about adding throttling in general (https://github.com/elastic/beats/issues/17775), but this may require many more changes.
In the case of Cloud Foundry we could add the rate limit per organization in the specific input.

Add a field or tag to the events to indicate that they are being throttled.

**Describe a specific use case for the enhancement or feature:**

On clusters with many organizations it may be good to limit the events rate so organizations make a fair use of resources.

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds a monitoring counter metric, `processors.rate_limit.n.dropped` where `n` is the the `n`th instance (1-based) of a `rate_limit` processor used in a Beat's configuration. This counter is incremented each time the processor drops an event due to rate limiting.

## Why is it important?

This allows users of the processor to understand whether their events are being rate limited and by how much.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Create a minimal Filebeat configuration with this processor in it.
   ```
   # filebeat.test.yml

   filebeat.inputs:
   - type: stdin

   processors:
     - rate_limit:
         limit: "1/m"

   output.console:
     enabled: true
     pretty: true

   http.enabled: true
   ```

2. Run Filebeat with the above configuration.
   ```
   filebeat -c filebeat.test.yml
   ```

3. Send events to Filebeat via STDIN at a rate faster than one event per minute (the rate limit).
4. In another window check that the Filebeat Stats API has the monitoring counter implemented by this PR and that it is incrementing as expected.
   ```
   curl -s 'http://localhost:5066/stats' | jq -c '.processor.rate_limit."1".dropped' 
   ```

## Related issues

- Closes #21020

### CODE REVIEW

File: libbeat/processors/ratelimit/docs/rate_limit.asciidoc
Comment (ycombinator) [inline libbeat/processors/ratelimit/docs/rate_limit.asciidoc:14]: This change is unrelated to this PR but since it's minor I thought I would include it in here. Let me know if you'd prefer to have it in its own PR.

File: libbeat/processors/ratelimit/rate_limit.go
Comment (ycombinator) [inline libbeat/processors/ratelimit/rate_limit.go:132]: I am on the fence about whether to reset this counter each time or let it grow so it's a true counter. I decided to not go with the latter since it would be more susceptible to wraparound issues. But happy to discuss and change!

File: libbeat/processors/ratelimit/rate_limit.go
Comment (ycombinator) [inline libbeat/processors/ratelimit/rate_limit.go:45]: I decided to keep track of the number of events that have been rate-limited (since the last time an event was allowed through; more on that in https://github.com/elastic/beats/pull/23330/files#r550323522). Alternatively we could simply have a boolean to indicate that rate limiting has happened. But I felt like a number might be more informative. Happy to discuss and change, though!

Comment (ycombinator) [reply libbeat/processors/ratelimit/rate_limit.go:45]: On a related note, it would be nice if we could include this metric as part of the libbeat monitoring infrastructure. For example, the `dns` processor is doing something like this. If you agree, I can implement this as part of this PR or a separate PR.

File: libbeat/processors/ratelimit/docs/rate_limit.asciidoc
Comment (jsoriano) [inline libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: I wonder if we should decide what field to use for this. This way users wouldn't need to think what field to use, with the risk of choosing a field that is used by something else. We could also provide a mapping for the field. And we could also keep this always enabled.

(Naming is complicated, I don't have a suggestion for this field :grimacing: )

As a example of something somehow similar, the `multiline` feature of `filebeat` adds a `multiline` flag in a known field (`log.flags`) when multiple lines are merged in a single event, and this cannot be disabled.

File: libbeat/processors/ratelimit/rate_limit.go
Comment (jsoriano) [inline libbeat/processors/ratelimit/rate_limit.go:132]: If multiple goroutines are sending events at the same time, all of them could see that `p.numRateLimited.Load() > 0`, and they could create multiple events with the same count before it is reset to 0. We could use the atomic swap to ensure that the counter is only set in one event.
```suggestion
	if p.config.MetricField != "" {
		if count := p.numRateLimited.Swap(0); count > 0 {
			event.PutValue(p.config.MetricField, count)
```

**Review by jsoriano — COMMENTED** [review]
I have added some thoughts, let me know what you think. Thanks for adding this!

File: libbeat/processors/ratelimit/docs/rate_limit.asciidoc
Comment (ycombinator) [reply libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: I recall having a conversation previously where we thought it might be a good idea to give the user control of the field: https://github.com/elastic/beats/issues/21020#issuecomment-742468061. But your points here about the advantages of having a fixed field are also valid. 

Maybe we should take a step back and ask what is the use case for including a field in the event that describes a past state, especially considering that events may be batched and retried by some outputs. I think the idea is to know that rate limiting is happening or not and, if it is, to what extent. So maybe a monitoring counter is the more reliable way of expressing this?

Comment (jsoriano) [reply libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: > I think the idea is to know that rate limiting is happening or not and, if it is, to what extent.

Yes, I think the same.

> So maybe a monitoring counter is the more reliable way of expressing this?

A monitoring counter sounds good, but I wonder if this is enough when multiple `rate_limit` processors are configured in the same beat, it can be difficult to see which one is being triggered. Though maybe this is not such a problem in real deployments.
Having a counter for each `rate_limit` could be an alternative, but it can complicate things, and I am not sure how each counter could be correlated to each config.
Having the info in the event is not so nice or reliable, but I guess it is easier to identify what kind of events are being throttled.

Maybe by now we can go on with a global monitoring counter (one for all `rate_limit` processors), and wait for feedback to see if this is enough. Having debug logging about the throttled events could also help with tricky cases and complicated configs.

Comment (ycombinator) [reply libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: > A monitoring counter sounds good, but I wonder if this is enough when multiple rate_limit processors are configured in the same beat, it can be difficult to see which one is being triggered. Though maybe this is not such a problem in real deployments.

I looked at how the `dns` processor solves this issue. It does so by giving each _instance_ of the processor it's own ID and then using it for logging and metrics monitoring purposes:

https://github.com/elastic/beats/blob/11c5367b33fb0a92d93c24fb6d382c1795aed5a5/libbeat/processors/dns/dns.go#L60-L65

So maybe we could play the same game here?

Alternatively we could create a new `processors.MonitoredProcessor` struct that implements `processors.Processor` but also implements a `MonitoringRegistry()` method that provides a monitoring registry instance for any processor that wishes to be monitored. This way we will have a consistent namespace for each processor instance within the libbeat monitoring registry for per-processor-instance metrics.

WDYT?

> Having debug logging about the throttled events could also help with tricky cases and complicated configs.
Agreed. We already have this today: https://github.com/elastic/beats/blob/11c5367b33fb0a92d93c24fb6d382c1795aed5a5/libbeat/processors/ratelimit/rate_limit.go#L89

Comment (jsoriano) [reply libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: Ok, if we have already other processors doing something like this I am ok with following the same approach :+1:

Comment (ycombinator) [reply libbeat/processors/ratelimit/docs/rate_limit.asciidoc:44]: Cool, I'm going to change this PR to remove the metric field on the event and instead setup a monitoring counter for the cumulative total number of events rate limited.

**Comment by ycombinator** [comment]
@jsoriano I've updated this PR significantly per https://github.com/elastic/beats/pull/23330#discussion_r553521910.  Please re-review when you get a chance.  Thanks!

File: libbeat/processors/ratelimit/rate_limit.go
Comment (jsoriano) [inline libbeat/processors/ratelimit/rate_limit.go:112]: Add the id to this debug message so it is easier to correlate with the monitoring counter?

File: CHANGELOG.next.asciidoc
Comment (jsoriano) [inline CHANGELOG.next.asciidoc:574]: Not added anymore :)
```suggestion
```

**Review by jsoriano — COMMENTED** [review]
Looks good, added some minor comments.

File: libbeat/processors/ratelimit/rate_limit.go
Comment (ycombinator) [reply libbeat/processors/ratelimit/rate_limit.go:112]: Yep, good catch. Will fix.

Comment (ycombinator) [reply libbeat/processors/ratelimit/rate_limit.go:112]: Actually the ID is already included by this line: https://github.com/elastic/beats/pull/23330/files#diff-67f3af920cbdfab0fed6f4a5f8077b14cd27cf91fdf20f53c3e02cac8e0a8b70R82. This is the same thing the `dns` processor is also doing for its logger: https://github.com/elastic/beats/blob/f5fd81729fc1dfab4e12a138893f9aec73921c75/libbeat/processors/dns/dns.go#L63


For example, this is what the log looks like (notice the `instance_id` at the end):

```
2021-01-13T11:20:55.263-0800    DEBUG   [processor.rate_limit]  ratelimit/rate_limit.go:112     event [&{2021-01-13 11:20:55.263477 -0800 PST m=+6.116778961 {} {"agent":{"ephemeral_id":"3b27d985-65b6-4591-9a21-5d53ca6a6721","id":"d8740c81-a061-490a-8c17-340ff37a95b3","name":"Shaunaks-MacBook-Pro-Work.local","type":"filebeat","version":"8.0.0"},"ecs":{"version":"1.6.0"},"host":"xxxxx","input":{"type":"stdin"},"log":{"file":{"path":""},"offset":0},"message":"bar"} {  false <nil>  4 0001-01-01 00:00:00 +0000 UTC 0s  map[] 0-0 } false}] dropped by rate_limit processor      {"instance_id": 1}
```

Comment (jsoriano) [reply libbeat/processors/ratelimit/rate_limit.go:112]: Oh ok, nice then :+1:

**Review by jsoriano — APPROVED** [review]

**Pull request merged**
